compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp

```diff
@@ -401,23 +401,71 @@ LogicalResult getStaticDims(Operation *op,
   return success();
 }
 
+/// Utility to get the static offsets, sizes and strides for
+/// `AIEX::NpuDmaMemcpyNdOp` with implicit addressing.
+LogicalResult getStaticDimsForImplicitAddressing(
+    Operation *op, MemRefType memrefType,
+    SmallVectorImpl<int64_t> &staticOffsets,
+    SmallVectorImpl<int64_t> &staticSizes,
+    SmallVectorImpl<int64_t> &staticStrides) {
+  // 1. Static offsets.
+  for (unsigned i = 0, n = staticOffsets.size(); i < n; i++) {
+    staticOffsets[i] = 0;
+  }
+  // 2. Static sizes.
+  SmallVector<int64_t> shapeArr;
+  for (auto shape : memrefType.getShape()) shapeArr.push_back(shape);
+  int64_t sizeIndex = staticSizes.size() - 1;
+  for (int64_t i = shapeArr.size() - 1; i >= 0; i--) {
+    staticSizes[sizeIndex--] = shapeArr[i];
+  }
+  for (int64_t i = sizeIndex; i >= 0; i--) {
+    staticSizes[i] = 0;
+  }
+  // 3. Static strides.
+  int64_t strideIndex = staticStrides.size() - 1;
+  // Since aiex.npu.dma_cpy_nd always has the 4-th dimension as 1, we store that
+  // in `prevStride`.
+  int64_t prevStride = 1;
+  for (int64_t i = shapeArr.size() - 1; i >= 0; i--) {
+    staticStrides[strideIndex] = prevStride;
+    prevStride = staticStrides[strideIndex--] * shapeArr[i];
+  }
+  for (int64_t i = strideIndex; i >= 0; i--) {
+    staticStrides[i] = prevStride;
+  }
+  return success();
+}
+
 /// Convert the `amdaie.npu.dma_cpy_nd` operation to `aiex.npu.dma_memcpy_nd`.
 LogicalResult npuDmaCpyNdOpToAIE(IRRewriter &rewriter,
                                  AMDAIE::NpuDmaCpyNdOp dmaOp,
                                  SmallVector<Operation *> &toBeErased,
                                  IRMapping &mapper, IRMapping &bindingsMapper) {
   rewriter.setInsertionPoint(dmaOp);
   // Convert bidirectional `amdaie.npu.dma_cpy_nd` op into two halves.
-  if (dmaOp.hasSourceAddressing()) {
+  if (dmaOp.hasSourceAddressing() || dmaOp.getSourceMemorySpaceAsUInt() == 0) {
```

Comment:
You're not really ensuring here that the addressing is on L3 side. This condition would still succeed if `dmaOp.hasSourceAddressing() == true` and `dmaOp.getSourceMemorySpaceAsUInt() != 0`, so it might be good to reflect that in the comment.